File: streamx-console/streamx-console-service/src/main/java/com/streamxhub/streamx/console/core/entity/FlinkVersion.java
/*
 *  Copyright (c) 2019 The StreamX Project
 *
 * <p>Licensed to the Apache Software Foundation (ASF) under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional information regarding
 * copyright ownership. The ASF licenses this file to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the License. You may obtain a
 * copy of the License at
 *
 * <p>http://www.apache.org/licenses/LICENSE-2.0
 *
 * <p>Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.streamxhub.streamx.console.core.entity;


import com.baomidou.mybatisplus.annotation.TableName;
import com.streamxhub.streamx.common.util.CommandUtils;
import com.streamxhub.streamx.common.util.DeflaterUtils;
import com.streamxhub.streamx.common.util.PropertiesUtils;
import lombok.Data;
import net.minidev.json.annotate.JsonIgnore;
import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.io.Serializable;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author benjobs
 */
@Data
@TableName("t_flink_version")
public class FlinkVersion implements Serializable {

    private Long id;

    private String flinkName;

    private String flinkHome;

    private String flinkConf;

    private String description;

    private String version;

    /**
     * 是否为默认版本.
     */
    private Boolean isDefault;

    private Date createTime;

    @JsonIgnore
    private transient final Pattern flinkVersionPattern = Pattern.compile("^Version: (.*), Commit ID: (.*)$");

    @JsonIgnore
    public String extractFlinkVersion(String path) {
        assert path != null;
        AtomicReference<String> flinkVersion = new AtomicReference<>();
        if (path != null) {
            String libPath = path.concat("/lib");
            File[] distJar = new File(libPath).listFiles(x -> x.getName().matches("flink-dist_.*\\.jar"));
            if (distJar == null || distJar.length == 0) {
                throw new IllegalArgumentException("[StreamX] can no found flink-dist jar in " + libPath);
            }
            if (distJar.length > 1) {
                throw new IllegalArgumentException("[StreamX] found multiple flink-dist jar in " + libPath);
            }
            List<String> cmd = Arrays.asList(
                "cd ".concat(path),
                String.format(
                    "java -classpath %s org.apache.flink.client.cli.CliFrontend --version",
                    distJar[0].getAbsolutePath()
                )
            );

            CommandUtils.execute(cmd, versionInfo -> {
                Matcher matcher = flinkVersionPattern.matcher(versionInfo);
                if (matcher.find()) {
                    flinkVersion.set(matcher.group(1));
                }
            });
        }
        return flinkVersion.get();
    }

    public void doSetFlinkConf() throws IOException {
        assert this.flinkHome != null;
        File yaml = new File(this.flinkHome.concat("/conf/flink-conf.yaml"));
        assert yaml.exists();
        String flinkConf = FileUtils.readFileToString(yaml);
        this.flinkConf = DeflaterUtils.zipString(flinkConf);
    }

    public void doSetVersion() throws IOException {
        assert this.flinkHome != null;
        this.setVersion(this.extractFlinkVersion(this.getFlinkHome()));
    }

    @JsonIgnore
    public Map<String, String> convertFlinkYamlAsMap() {
        String flinkYamlString = DeflaterUtils.unzipString(flinkConf);
        return PropertiesUtils.loadFlinkConfYaml(flinkYamlString);
    }

    public void unzipFlinkConf() {
        this.flinkConf = DeflaterUtils.unzipString(this.flinkConf);
    }

    public String getLargeVersion() {
        return this.version.substring(0, this.version.lastIndexOf("."));
    }

    public String getVersionOfFirst() {
        return this.version.split("\\.")[0];
    }

    public String getVersionOfMiddle() {
        return this.version.split("\\.")[1];
    }

    public String getVersionOfLast() {
        return this.version.split("\\.")[2];
    }

}

File: streamx-common/src/main/scala/com/streamxhub/streamx/common/domain/FlinkVersion.scala
/*
 * Copyright (c) 2019 The StreamX Project
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.streamxhub.streamx.common.domain

import com.streamxhub.streamx.common.util.CommandUtils
import org.apache.commons.lang3.StringUtils

import java.io.File
import java.net.{URL => NetURL}
import java.util.concurrent.atomic.AtomicReference
import java.util.function.Consumer
import java.util.regex.Pattern
import scala.collection.JavaConversions._

/**
 * @param flinkHome actual flink home that must be a readable local path
 * @author benjobs
 */
class FlinkVersion(val flinkHome: String) extends java.io.Serializable {

  private[this] lazy val FLINK_VER_PATTERN = Pattern.compile("^(\\d+\\.\\d+)(\\.)?.*$")

  private[this] lazy val FLINK_VERSION_PATTERN = Pattern.compile("^Version: (.*), Commit ID: (.*)$")

  private[this] lazy val FLINK_SCALA_VERSION_PATTERN = Pattern.compile("^flink-dist_(.*)-(.*).jar$")

  lazy val scalaVersion: String = {
    val matcher = FLINK_SCALA_VERSION_PATTERN.matcher(flinkDistJar.getName)
    matcher.matches()
    matcher.group(1)
  }

  lazy val fullVersion: String = s"${version}_$scalaVersion"

  lazy val flinkLib: File = {
    require(flinkHome != null, "[StreamX] flinkHome must not be null.")
    require(new File(flinkHome).exists(), "[StreamX] flinkHome must be exists.")
    val lib = new File(s"$flinkHome/lib")
    require(lib.exists() && lib.isDirectory, s"[StreamX] $flinkHome/lib must be exists and must be directory.")
    lib
  }

  lazy val flinkLibs: List[NetURL] = flinkLib.listFiles().map(_.toURI.toURL).toList

  lazy val version: String = {
    val flinkVersion = new AtomicReference[String]
    val cmd = List(
      s"cd ${flinkLib.getAbsolutePath}",
      s"java -classpath ${flinkDistJar.getAbsolutePath} org.apache.flink.client.cli.CliFrontend --version"
    )
    CommandUtils.execute(cmd, new Consumer[String]() {
      override def accept(out: String): Unit = {
        val matcher = FLINK_VERSION_PATTERN.matcher(out)
        if (matcher.find) {
          flinkVersion.set(matcher.group(1))
        }
      }
    })
    val version = flinkVersion.get
    doCheckVersion(version)
    version
  }

  def doCheckVersion(version: String): Unit = {
    if (StringUtils.isEmpty(version)) {
      throw new IllegalStateException("[StreamX] parse flink version failed.")
    }
    if (version.split("\\.").length != 3) {
      throw new IllegalStateException("[StreamX] parse illegal version.")
    }
  }

  // flink major version, like "1.13", "1.14"
  lazy val majorVersion: String = {
    if (version == null) {
      null
    } else {
      val matcher = FLINK_VER_PATTERN.matcher(version)
      matcher.matches()
      matcher.group(1)
    }
  }

  lazy val flinkDistJar: File = {
    val distJar = flinkLib.listFiles().filter(_.getName.matches("flink-dist_.*\\.jar"))
    distJar match {
      case x if x.isEmpty =>
        throw new IllegalArgumentException(s"[StreamX] can no found flink-dist jar in $flinkLib")
      case x if x.length > 1 =>
        throw new IllegalArgumentException(s"[StreamX] found multiple flink-dist jar in $flinkLib")
      case _ =>
    }
    distJar.head
  }

  // streamx flink shims version, like "streamx-flink-shims_flink-1.13"
  lazy val shimsVersion: String = s"streamx-flink-shims_flink-$majorVersion"

  override def toString: String =
    s"""
       |----------------------------------------- flink version -----------------------------------
       |     flinkHome    : $flinkHome
       |     distJarName  : ${flinkDistJar.getName}
       |     flinkVersion : $version
       |     majorVersion : $majorVersion
       |     scalaVersion : $scalaVersion
       |     shimsVersion : $shimsVersion
       |-------------------------------------------------------------------------------------------
       |""".stripMargin

}
